### PR TITLE
Allow git branches with file-system-invalid chars

### LINKED ISSUE #1004: Make sure that "interesting" branch names are escaped
When cloning git repos we put them in the `.spago` folder, with the ref being part of the path. This is all fine, except that branch names can contain slashes, with would mess up the path.

[spago-legacy](https://github.com/purescript/spago-legacy/blob/11e68ece50ede286660452ef370baa3ab5c33c02/test/UnitSpec.hs#L96) had tests for this, and we [have a commented out test](https://github.com/purescript/spago/blob/32d89b515e656a57ee78552b87f31cf93d8decdc/test/Spago/Install.purs#L104) (even though we should have the QuickCheck ones too) that should check for this.

The fix should be something along the lines [of the old implementation that we had](https://github.com/purescript/spago-legacy/blob/11e68ece50ede286660452ef370baa3ab5c33c02/src/Spago/FetchPackage.hs#L219)

### PR DESCRIPTION
**Pull request by JordanMartinez**
### Description of the change

Fixes #1004

### Checklist:

- [ ] Added the change to the "Unreleased" section of the changelog
- [ ] Added some example of the new feature to the `README`
- [ ] Added a test for the contribution (if applicable)

**P.S.**: the above checks are not compulsory to get a change merged, so you may skip them. However, taking care of them will result in less work for the maintainers and will be much appreciated 😊

### CODE REVIEW

File: test/Spago/Install.purs
Comment (JordanMartinez) [inline test/Spago/Install.purs:132]: I'm not sure what `installs-with-slash.txt` is supposed to say, but if this succeeds, is that good enough?

File: src/Spago/Config.purs
Comment (f-f) [inline src/Spago/Config.purs:38]: We already import this as `String`

File: src/Spago/Config.purs
Comment (f-f) [inline src/Spago/Config.purs:31]: Could we qualify all of these function imports? We generally import directly only types, everything else is qualified.

File: test/Spago/Install.purs
Comment (f-f) [reply test/Spago/Install.purs:132]: I think that's fine, but we should also look that the folder in the `.spago` folder has the name that we expect it to have

**Comment by f-f** [comment]
The issue also mentions Quick check tests that check for the injectivity of the naming function, though I don't have a link at hand

**Comment by JordanMartinez** [comment]
> The issue also mentions Quick check tests that check for the injectivity of the naming function, though I don't have a link at hand

Looks like this one? https://github.com/purescript/spago-legacy/blob/master/test/UnitSpec.hs

**Comment by f-f** [comment]
Ah yes! It's this one. Not sure if it's worth porting since the implementation has been stable for years, so we don't have to necessarily have it in the PR

**Comment by JordanMartinez** [comment]
> Ah yes! It's this one. Not sure if it's worth porting since the implementation has been stable for years, so we don't have to necessarily have it in the PR

I think it's worth porting.

**Comment by f-f** [comment]
@JordanMartinez I think this one is good to go once the conflicts are sorted out

**Comment by JordanMartinez** [comment]
Can this get an approval?

**Review by f-f — APPROVED** [review]
Thanks!

**Pull request merged**
